Allow installing Gradle API jars into a local repository

wolfs opened the pull request.

wolfs committed:
First try at creating Repository for Gradle API

by downloading the bin distribution and publishing it.

wolfs committed:
Use an artifact transform for the jars list

wolfs committed:
Add README for gradle-repository

lptr commented on gradle-repository/.gitattributes:
```diff
@@ -0,0 +1,6 @@
+#
+# These are explicitly windows files and should use crlf
+*.bat           text eol=crlf
+
```
Are separate `.gitattributes` and `.gitignore` necessary here?

wolfs replied:
I removed the `.gitignore`, since the files are already covered by the parent one. The `.gitattributes` file is necessary to keep the line endings for `gradlew.bat`.

lptr replied:
This is nit-picking, so ignore it if you want, but shouldn't the `.gitattributes` file go in the root of the repo then?

asodja commented on gradle-repository/buildSrc/src/main/kotlin/gradlebuild.gradle-api-repository.gradle.kts:
```diff
@@ -0,0 +1,126 @@
+    into(project.buildDir.resolve("libs"))
+}
+
+afterEvaluate {
```
Why is `afterEvaluate` needed here?

asodja replied:
So I see publishing needs that. What part of `publishing` doesn't work here?

wolfs replied:
I left a comment explaining why we need the `afterEvaluate`.

wolfs committed:
Publish a fat Gradle API jar

instead of the individual JARs. That simplifies the publishing
configuration, since we only have one publication now.

wolfs committed:
Extract artifactType for gradle-libs-dir

wolfs committed:
Remove .gitignore

The parent .gitignore covers everything already.

wolfs commented:
@lptr @asodja I changed the logic so we only publish a fat Gradle API jar instead of individual JARs. That simplifies the logic for publishing quite a bit. PTAL!

asodja reviewed (approved):
LGTM

lptr commented on gradle-repository/buildSrc/src/main/kotlin/ExplodeZipAndFindJars.kt:
```diff
@@ -0,0 +1,37 @@
+                var shortName: String = zipEntry.name
+                if (shortName.contains('/')) {
+                    shortName = shortName.substring(shortName.lastIndexOf('/') + 1)
+                }
```
It would be good to capture the fact that this also flattens the output in the name. From `ExplodeZipAndFindJars` I expected the filtering, but not the flattening.

wolfs replied:
Would `ExtractJarsFromZip` be better? Do you have a suggestion?

lptr replied:
Sounds good.

wolfs replied:
Renamed.

lptr commented on gradle-repository/buildSrc/src/main/kotlin/gradlebuild.gradle-api-repository.gradle.kts:
```diff
@@ -0,0 +1,85 @@
+
+gradleApiRepository.apply {
+    gradleVersion.convention(
+        providers.gradleProperty("gradleApiVersion").orElse("7.4.1")
```
Why `7.4.1` specifically? Why not the current Gradle version?

wolfs replied:
You mean the current version from the wrapper? I don't feel strongly what the convention should be. Do you think it makes sense to force the user to provide a version via a Gradle property and not have any default?

lptr replied:
Forcing the user sounds like the right way to go here, no? Does it mean anything to say "publish a Gradle API so I can compile against it, I don't care which version?"

wolfs replied:
I removed the default.

lptr commented on gradle-repository/buildSrc/src/main/kotlin/gradlebuild.gradle-api-repository.gradle.kts:
```diff
@@ -0,0 +1,85 @@
+            url = uri("https://services.gradle.org")
+            patternLayout {
+                artifact("/${distUrl}/[module]-[revision]-bin(.[ext])")
+            }
```
Does using an Ivy layout have some benefits here? If not, wouldn't a Maven one be simpler?

wolfs replied:
Ivy layout in this case is a workaround for "Use an arbitrary URL and cache the result". You can't use a Maven layout, since a Maven layout can't be customized.

lptr replied:
Why do we need to customize it?

lptr commented on gradle-repository/buildSrc/src/main/kotlin/gradlebuild.gradle-api-repository.gradle.kts:
```diff
@@ -0,0 +1,85 @@
+afterEvaluate {
+
+    // We need to configure publishing in afterEvaluate,
+    // since the repository URL and the MavenPublication.version aren't providers.
```
```suggestion
    // since the repository URL and the MavenPublication.version aren't providers yet.
```

wolfs committed:
Improved afterEvaluateDocumentation

Co-authored-by: Lóránt Pintér <lorant.pinter@gmail.com>

wolfs committed:
Rename ExplodeZipAndFindJars -> ExtractJarsFromZip

wolfs committed:
Remove default for Gradle API version

The user now needs to specify the version on the command line.
It doesn't make much sense to have a default here.

lptr reviewed (approved):
LGTM
